js/msg.js

```diff
@@ -1,258 +1,168 @@
+      if (!msg.isIgnorableError(err)) {
+        console.warn(err);
+      }
+    },
 
-  function onTab(fn) {
-    initHandler();
-    handler.tab.push(fn);
-  }
+    on(fn) {
+      handler.both.add(fn);
+    },
 
-  function onExtension(fn) {
-    initHandler();
-    handler.extension.push(fn);
-  }
+    onTab(fn) {
+      handler.tab.add(fn);
+    },
 
-  function off(fn) {
-    for (const type of ['both', 'tab', 'extension']) {
-      const index = handler[type].indexOf(fn);
-      if (index >= 0) {
-        handler[type].splice(index, 1);
+    onExtension(fn) {
+      handler.extension.add(fn);
+    },
+
+    off(fn) {
+      for (const type of TARGETS.all) {
+        handler[type].delete(fn);
       }
-    }
+    },
+
+    send(data, target = 'extension') {
+      return browser.runtime.sendMessage({data, target})
+        .then(unwrapData);
+    },
+
+    sendTab(tabId, data, options, target = 'tab') {
+      return browser.tabs.sendMessage(tabId, {data, target}, options)
+        .then(unwrapData);
+    },
+
+    _execute(types, ...args) {
+      let result;
+      for (const type of types) {
+        for (const fn of handler[type]) {
+          let res;
+          try {
+            res = fn(...args);
+          } catch (err) {
+            res = Promise.reject(err);
+          }
+          if (res !== undefined && result === undefined) {
+            result = res;
+          }
+        }
+      }
+      return result;
+    },
+  };
+
+  function getExtBg() {
+    const fn = chrome.extension.getBackgroundPage;
+    return fn && fn();
   }
 
-  function initHandler() {
-    if (handler) {
-      return;
-    }
-    handler = {
-      both: [],
-      tab: [],
-      extension: []
-    };
-    if (isBg) {
-      window._msg.handler = handler;
-    }
-    chrome.runtime.onMessage.addListener(handleMessage);
+  function getExtBgIfReady() {
+    const bg = getExtBg();
+    return bg && bg.document && bg.document.readyState !== 'loading' && bg;
   }
 
-  function executeCallbacks(callbacks, ...args) {
-    let result;
-    for (const fn of callbacks) {
-      const data = withPromiseError(fn, ...args);
-      if (data !== undefined && result === undefined) {
-        result = data;
-      }
-    }
-    return result;
+  function getRuntimeBg() {
+    return new Promise(resolve =>
+      chrome.runtime.getBackgroundPage(bg =>
+        resolve(!chrome.runtime.lastError && bg)));
   }
 
-  function handleMessage(message, sender, sendResponse) {
-    const handlers = message.target === 'tab' ?
-      handler.tab.concat(handler.both) : message.target === 'extension' ?
-      handler.extension.concat(handler.both) :
-      handler.both.concat(handler.extension, handler.tab);
-    if (!handlers.length) {
-      return;
+  function handleMessage({data, target}, sender, sendResponse) {
+    const res = msg._execute(TARGETS[target] || TARGETS.all, data, sender);
+    if (res instanceof Promise) {
+      handleResponseAsync(res, sendResponse);
```

Comment:
How about renaming it to `wrapData` function (compare to `unwrapData`) and making it returns the message object so we can use it like:
```js
wrapData(res).then(sendResponse);
```